File: Model/detector_parity.js
// Full-detector parity (JS side): run the on-device detector and confirm its
// verdicts match the backend's. Run AFTER detector_parity.py.
const fs = require("fs");
const path = require("path");

const EXT = path.join(__dirname, "..", "Extension");
require(path.join(EXT, "heuristics.js"));
require(path.join(EXT, "tier1.js"));
const Detector = require(path.join(EXT, "detector.js"));
const model = JSON.parse(fs.readFileSync(path.join(EXT, "model.json"), "utf-8"));
const data = JSON.parse(fs.readFileSync(path.join(__dirname, "detector_parity_data.json"), "utf-8"));

const det = Detector.make(model);

let fails = 0;
console.log("risk(py/js) | score(py/js) | ml(py/js)        | text");
console.log("-".repeat(78));
for (const row of data.rows) {
  const r = det.detect(row.text, "");
  const riskOk = r.risk === row.risk;
  const scoreOk = r.score === row.score;
  const mlOk = Math.abs((r.ml_prob ?? 0) - (row.ml ?? 0)) <= 2e-3;
  const ok = riskOk && scoreOk && mlOk;
  if (!ok) fails++;
  console.log(
    `${(ok ? "OK " : "XX ")}${row.risk}/${r.risk} | ${row.score}/${r.score} | ` +
    `${row.ml}/${r.ml_prob} | ${row.text.slice(0, 30)}`
  );
}
console.log("-".repeat(78));
console.log(fails === 0 ? "DETECTOR PARITY OK — on-device == backend" : `FAILED — ${fails} mismatches`);
process.exit(fails === 0 ? 0 : 1);

File: Model/detector_parity.py
# Full-detector parity (Python side): for each message, compute the backend's
# fused verdict exactly as /scan would, and dump it for the JS side to match.
import json
import sys
from pathlib import Path

if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")

sys.path.insert(0, str(Path(__file__).parent.parent / "Backend"))
from heuristics import score_text          # noqa: E402
import classifier                          # noqa: E402
from main import fuse                      # noqa: E402

from parity_test import MESSAGES           # reuse the same message set  # noqa: E402

rows = []
for m in MESSAGES:
    h = score_text(m)
    ml = classifier.predict_proba(m)
    risk, _ = fuse(h, ml)
    rows.append({
        "text": m, "risk": risk, "score": h.score,
        "ml": round(ml, 3) if ml is not None else None,
    })

(Path(__file__).parent / "detector_parity_data.json").write_text(
    json.dumps({"rows": rows}, ensure_ascii=False, indent=1), encoding="utf-8")
print(f"Wrote backend verdicts for {len(rows)} messages -> detector_parity_data.json")
